core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -920,36 +935,72 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       auditContext.setSrcInode(inode);
-
       List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        try (TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath)) {
-          try {
-            mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-          } catch (AccessControlException e) {
-            auditContext.setAllowed(false);
-            throw e;
-          }
-          for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-            child.lockReadAndCheckParent(inode);
-            try {
-              // the path to child for getPath should already be locked.
-              tempInodePath.setDescendant(child, mInodeTree.getPath(child));
-              ret.add(getFileInfoInternal(tempInodePath));
-            } finally {
-              child.unlockRead();
-            }
-          }
-        }
-      } else {
-        ret.add(getFileInfoInternal(inodePath));
+      DescendantType descendantTypeForListStatus = (listStatusOptions.isRecursive())
+          ? DescendantType.ALL : DescendantType.ONE;
+      listStatusInternal(inodePath, auditContext, descendantTypeForListStatus, ret);
+
+      // If we are listing the status of a directory, we remove the directory info that we inserted
+      if (inode.isDirectory() && ret.size() >= 1) {
+        ret.remove(ret.size() - 1);
       }
+
       auditContext.setSucceeded(true);
       Metrics.FILE_INFOS_GOT.inc();
       return ret;
     }
   }
 
+  /**
+   * Lists the status of the path in {@link LockedInodePath}, possibly recursively depending on
+   * the descendantType. The result is returned via a list specified by statusList, in postorder
+   * traversal order.
+   *
+   * @param currInodePath the inode path to find the status
+   * @param auditContext the audit context to return any access exceptions
+   * @param descendantType if the currInodePath is a directory, how many levels of its descendant
+   *                       should be returned
+   * @param statusList To be populated with the status of the files and directories requested
+   * @throws AccessControlException if the path can not be read by the user
+   * @throws FileDoesNotExistException if the path does not exist
+   * @throws UnavailableException if the service is temporarily unavailable
+   */
+  private void listStatusInternal(LockedInodePath currInodePath,
+                                  AuditContext auditContext,
+                                  DescendantType descendantType,
+                                  List<FileInfo> statusList)
+      throws FileDoesNotExistException, UnavailableException, AccessControlException {
+    Inode<?> inode = currInodePath.getInode();
+    if (inode.isDirectory()) {
+      try {
+        mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, currInodePath);
+      } catch (AccessControlException e) {
+        auditContext.setAllowed(false);
+        if (descendantType == DescendantType.ALL) {
```

Comment:
The posix behavior shows you the entries you have access to and reports a permission denied error message for what you don't. I think we should do something similar instead of silently ignoring

Reply:
I'm not sure what the best way to do this is, since it looks like we want to return results back, but also return exception/error messages. I don't think we have a mechanism to do both.

Reply:
maybe we can create a subclass of `FileInfo` called `ErrorFileInfo` that wraps the exception? and return that?

Reply:
Added a TODO for this item